Decode and verify UAA JWT tokens without connecting to UAA server

hairyhum opened the pull request:
Fixes #3
Uses https://github.com/rabbitmq/uaa_jwt library to decode a token and verify a signature.
Signing keys should be predefined in the uaa_jwt application environment

Commit:
Decode and verify UAA JWT tokens without connecting to UAA server

Fixes #3
Uses rabbitmq/uaa_jwt library to decode a token and verify signature.
Signing keys should be predefined in the uaa_jwt application environment

hairyhum commented:
Using https://github.com/rabbitmq/rabbitmq-common/pull/161 packages Elixir dependencies and Elixir itself into `*.ez` archives.

Commit:
Fix test to represent Jwt decoded extended scope permissions

hairyhum commented:
Rebased from `master`

dumbbell committed:
Makefile: Do not depend twice on `rabbit`

[#136341089]

dumbbell committed:
Makefile: Add rabbit_common to BUILD_DEPS

It's required for the Erlang.mk plugins.

dumbbell committed:
Remove build.config; we use a vanilla erlang.mk

dumbbell committed:
Makefile: amqp_client is not needed, remove it from DEPS

kjnilsson commented on src/rabbit_auth_backend_uaa.erl:
```diff
@@ -84,61 +88,25 @@ with_token(Token, Fun) ->
-                _   -> {error, {Code, Body}}
-            end;
-        {error, _} = E -> E
+    case 'Elixir.UaaJWT':decode_and_verify(Token) of
```
Do we need to decode and verify the token each time? How about storing the decoded claims map in `impl` instead?

kjnilsson commented on src/rabbit_auth_backend_uaa.erl:
```diff
@@ -23,7 +23,8 @@
+         check_topic_access/4]).
 
 %% httpc seems to get racy when using HTTP 1.1
 -define(HTTPC_OPTS, [{version, "HTTP/1.0"}]).
```
is this used? likewise `description()` below refers to HTTP

Commit:
Command to add UAA signing keys

Commit:
Test command validation

Commit:
Test key validation when adding via cli command

Commit:
Test default key support

Commit:
Do not decode token every time permission is checked.

Decoded token is saved to `impl`.
When permission is checked, the `exp` field of the token
is compared to system_time and if the token is expired
`{error_message, "Token expired"}` is returned.

Commit:
Test token expiration

kjnilsson commented on src/rabbit_auth_backend_uaa.erl:
```diff
@@ -55,90 +53,61 @@ user_login_authorization(Username) ->
-                    Scopes = get_scopes(UserData),
-                    rabbit_oauth2_scope:topic_access(Resource, Permission, Context, Scopes)
-               end).
+check_vhost_access(#auth_user{username = Username, impl = DecodedToken},
```
`Username` is no longer used.

Commit:
Return error instead of error_message to comply with authz_backend API

kjnilsson reviewed (changes requested):
`make tests` fail unless `make dist` is first run.

Commit:
Handle json parsing error

hairyhum commented:
`test-dist` now is built before tests are run:
https://github.com/rabbitmq/rabbitmq-common/commit/60f5bb2579735cc74cddfd8cfe70fc980c85faa5

kjnilsson committed:
Add cowlib to DEPS

kjnilsson committed:
README clarifications

Commit:
Improve tests

Commit:
Scope examples in README
